### ISSUE TITLE
Segfault when calling a PHP built-in function

### ISSUE THREAD

**Issue by z80maniac** [CopernicaMarketingSoftware/PHP-CPP#421]
When an extension calls any built-in PHP function, it crashes, i.e. this code crashes:
```cpp
Php::call("var_export", "function call");
```

Note: it only happens when PHP-CPP is built in release mode (with ```make release```).

Full example source code:
```cpp
#include <phpcpp.h>
#include <iostream>

void myFunction() {
    Php::call("var_export", "function call");
    Php::out << std::endl << "php out" << std::endl;
}

extern "C" PHPCPP_EXPORT void *get_module() {
    static Php::Extension extension("my_extension", "1.0");
    extension.add<myFunction>("myFunction");
    return extension;
}
```

Build:
```sh
g++ -shared -fPIC -o crashtest.so crashtest.cpp -lphpcpp
```

Test:
```sh
php -dextension=./crashtest.so -r 'myFunction();'
```

Output:
```
'function call'Segmentation fault (core dumped)
```

As mentioned above, the crash only happens when PHP-CPP is built in release mode. When PHP-CPP is built in debug mode, the output is as follows:
```
'function call'
php out
```
Which is the expected output.

Note: the crash does not happen in ```Php::out```. You can remove it and the extension will still crash. I just added it to check if the test reaches that point in code.

Tested on a clean Ubuntu 18.04 with PHP v7.2.

**Comment by scorninpc** [comment]
Works fine for me

![Seleção_099](https://user-images.githubusercontent.com/2607849/59791874-67cb0f00-92a9-11e9-8cd0-a058ad54086d.png)

What did you mean with release and debug mode? -g flag, what?

**Comment by z80maniac** [comment]
I mean, the PHP-CPP was built with ```make release```, not just ```make```.

**Comment by scorninpc** [comment]
Same, no problem

![Seleção_108](https://user-images.githubusercontent.com/2607849/59797200-90f19c80-92b5-11e9-9dd9-66539c708849.png)

release parameter add  -02 flag only, nothing more them tell to compiler to optimize the code

**Comment by z80maniac** [comment]
Here's a fully self-contained example that can be run via Docker:
```bash
#!/usr/bin/env bash
set -e
cd "$(dirname -- "$(readlink -f -- "$0")")"

apt update
DEBIAN_FRONTEND=noninteractive apt install -yq php-dev git

mkdir ../build
cd ../build

git clone https://github.com/CopernicaMarketingSoftware/PHP-CPP.git
pushd PHP-CPP && make release && make install && popd

cat << EOF > crashtest.cpp
#include <phpcpp.h>
#include <iostream>

void myFunction() {
    Php::call("var_export", "function call");
    Php::out << std::endl << "php out" << std::endl;
}

extern "C" PHPCPP_EXPORT void *get_module() {
    static Php::Extension extension("my_extension", "1.0");
    extension.add<myFunction>("myFunction");
    return extension;
}
EOF

g++ -shared -fPIC -o crashtest.so crashtest.cpp -lphpcpp

echo "Starting the test..."
php -dextension=./crashtest.so -r 'myFunction();'
echo "The test has been passed."
```

Name this file ```run.sh``` and run Docker in the same directory:
```sh
docker run --rm -v"$(pwd)":/test ubuntu:18.04 bash test/run.sh
```

The last lines will be:
```
Starting the test...
'function call'test/run.sh: line 33: 10602 Segmentation fault      (core dumped) php -dextension=./crashtest.so -r 'myFunction();'
```

Now, if you remove the ```release```, the test will be passed correctly:
```
Starting the test...
'function call'
php out
The test has been passed.
```

Maybe it's a problem with PHP 7.2 only, and not with PHP 7.3.

**Comment by scorninpc** [comment]
Yes, can be the 7.2 problem, but as I saw, release only optimize the code, there is no way can be difference.

Sorry for I cannot test docker test

If you see Makefile, you can run in production with "make" only

**Comment by z80maniac** [comment]
> can be the 7.2 problem

In fact, it may be Ubuntu problem. I ran the ```run.sh``` against ```php:7.3.6```, ```php:7.2.19``` and ```php:7.1.30``` docker images - and they all passed the test. But when I run it in Ubuntu, it crashes. And the crash is reproducible everywhere: in Docker, in LXC and on my host machine with Kubuntu 19.04.

So, may be it's not a problem in PHP-CPP. Though, I can't say for sure.

> there is no way can be difference

There can be a difference if a code is written incorrectly (e.g. produces undefined behavior according to the standard). I encountered this situation many times before in different projects.

> Sorry for I cannot test docker test

OK. I just wanted to show that the problem is really not on my side. Instead of Docker, you can try LXC, virtual machine or a live disk - anything that gives you a clean Ubuntu install.

> you can run in production with "make" only

As a last resort, yes. But I found out that if PHP-CPP is compiled inside a ```php:*``` Docker container, then it can be used to compile extensions that don't crash on Ubuntu. So, there's a semi-decent workaround at least.

**Comment by scorninpc** [comment]
What the exactly php7.2 version do you have?

**Comment by z80maniac** [comment]
On my host machine (Kubuntu 19.04):
```
$ php -v
PHP 7.2.19-0ubuntu0.19.04.1 (cli) (built: Jun  4 2019 14:44:42) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.2.0, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.2.19-0ubuntu0.19.04.1, Copyright (c) 1999-2018, by Zend Technologies
```

In LXC and Docker containers (Ubuntu 18.04):
```
# php -v
PHP 7.2.19-0ubuntu0.18.04.1 (cli) (built: Jun  4 2019 14:48:12) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.2.0, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.2.19-0ubuntu0.18.04.1, Copyright (c) 1999-2018, by Zend Technologies
```

However, as I said earlier, the test is passed successfully on ```php:7.2.19``` Docker image, which is Debian-based (Stretch).
